src/yggdrasil/packetconn.go

```diff
@@ -0,0 +1,277 @@
+package yggdrasil
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Arceliar/phony"
+	"github.com/yggdrasil-network/yggdrasil-go/src/crypto"
+	"github.com/yggdrasil-network/yggdrasil-go/src/types"
+)
+
+type packet struct {
+	addr    net.Addr
+	payload []byte
+}
+
+type PacketConn struct {
+	phony.Inbox
+	net.PacketConn
+	sessions      *sessions
+	closed        bool
+	readCallback  func(net.Addr, []byte)
+	readBuffer    chan packet
+	readDeadline  *time.Time
+	writeDeadline *time.Time
+}
+
+func newPacketConn(ss *sessions) *PacketConn {
+	return &PacketConn{
+		sessions:   ss,
+		readBuffer: make(chan packet),
+	}
+}
+
+func (c *PacketConn) _sendToReader(addr net.Addr, payload []byte) {
+	if c.readCallback == nil {
+		c.readBuffer <- packet{
```

Comment:
This should probably be in a `select` statement with a default case that drops the packet, otherwise forgetting to read would deadlock the whole node.